### PR TITLE
Support loading benchmarks truly from classpath

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 1]
```diff
@@ -277,11 +276,21 @@ private String benchmarkName(Path benchmarkFile)
     {
         Optional<List<String>> activeBenchmarks = properties.getActiveBenchmarks();
         if (activeBenchmarks.isPresent()) {
-            return pathContainsAny(activeBenchmarks.get());
+            return benchmarkNameIn(activeBenchmarks.get());
         }
         return path -> true;
     }
 
+    private Predicate<Path> benchmarkNameIn(List<String> activeBenchmarks)
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: static

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 2]
```diff
@@ -272,23 +305,13 @@ private String benchmarkName(Path benchmarkFile)
...
-    {
-        Optional<List<String>> activeBenchmarks = properties.getActiveBenchmarks();
-        if (activeBenchmarks.isPresent()) {
-            return benchmarkNameIn(activeBenchmarks.get());
-        }
-        return path -> true;
-    }
-
-    private Predicate<Path> benchmarkNameIn(List<String> activeBenchmarks)
+    private List<String> activeBenchmarkPatterns()
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: It looks like these changes should belong to different commit

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 3]
```diff
@@ -353,4 +376,18 @@ private String createFormatString(Collection<Benchmark> benchmarks)
         int indent = 3;
         return "\t| %-" + (nameMaxLength + indent) + "s | %-" + (dataSourceMaxLength + indent) + "s | %-4s | %-8s | %-11s |";
     }
+
+    private static ByteSource asByteSource(Resource resource)
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: maybe put into into `ResourceUtills`?

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java [hunk 4]
```diff
@@ -122,6 +128,56 @@ public void shouldLoadSimpleBenchmark()
...
+    }
+
+    private void checkBenchmarksLoading(String benchmarksDir)
+    {
+        withBenchmarksDir(benchmarksDir);
+
+        // When
+        List<Benchmark> benchmarks = loader.loadBenchmarks("sequenceId");
+
+        // Then
```
Comment (kokosing) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: do we need such comments?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 5]
```diff
@@ -139,32 +137,65 @@
...
+                .flatMap(pattern -> getResources(patternResolver, benchmarksDir + "/" + pattern))
+                .peek(resource -> LOGGER.info("Benchmark found: {}", resource))
                 .collect(toList());
-        benchmarkFiles.stream().forEach((path) -> LOGGER.info("Benchmark found: {}", path.toString()));
 
         return benchmarkFiles;
     }
 
-    private List<Benchmark> loadBenchmarks(String sequenceId, List<Path> benchmarkFiles)
+    private ResourcePatternResolver getResourcePatternResolver()
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: static?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 6]
```diff
@@ -139,32 +137,65 @@
...
+    {
+        FileSystemResourceLoader resourceLoader = new FileSystemResourceLoader()
+        {
+            @Override
+            protected Resource getResourceByPath(String path)
+            {
+                if (path != null && path.startsWith("/")) {
+                    // Workaround superclass's stripping of leading '/'. Shameful.
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: remove `Shameful.`

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 7]
```diff
@@ -139,32 +137,65 @@
...
                 .collect(toList());
-        benchmarkFiles.stream().forEach((path) -> LOGGER.info("Benchmark found: {}", path.toString()));
 
         return benchmarkFiles;
     }
 
-    private List<Benchmark> loadBenchmarks(String sequenceId, List<Path> benchmarkFiles)
+    private ResourcePatternResolver getResourcePatternResolver()
+    {
+        FileSystemResourceLoader resourceLoader = new FileSystemResourceLoader()
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Can you please extract this as separate named class and then please add javadoc which explains why do we need to override it?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 6 above)
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: If understand correctly if path is `/dir` to make `//dir` so the underlying class can remove one `\`, am I right?

What happens where there is no `/` in the path?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 8]
```diff
@@ -139,32 +137,65 @@
...
+                }
+                return super.getResourceByPath(path);
+            }
+        };
+
+        resourceLoader.setClassLoader(Thread.currentThread().getContextClassLoader());
+        return new PathMatchingResourcePatternResolver(resourceLoader);
+    }
+
+    private static Stream<Resource> getResources(ResourcePatternResolver patternResolver, String locationPattern)
```
Comment (kokosing) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Generally I think you could wrap usage of `ResourcePR` as separate class where you could expose only methods which should be used from here. Then you new class API could not to have any `throws` clause. 

IMO it would be nice to do this that way as this class would be less polluted with different problems. Would be more focuses on its main responsibility and not tools it is using.

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java [hunk 9]
```diff
@@ -271,12 +271,13 @@ public void getAllBenchmarks_activeVariables_with_regex()
...
         assertLoadedBenchmarksCount(4).forEach(benchmark ->
-                        assertThat(benchmark.getVariables().get("format")).isIn("orc", "txt")
+                assertThat(benchmark.getVariables().get("format")).isIn("orc", "txt")
         );
     }
 
     @Test
-    public void allBenchmarks_load_only_not_executed_within_two_days() {
+    public void allBenchmarks_load_only_not_executed_within_two_days()
+    {
```
Comment (kokosing) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: why is so? 

I think we need to have a codestyle file added to this project and make sure that codestyle is checked during the build.

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 2 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Not really. The semantics didn't change, the mechanics did, as the pattern is now evaluated during scanning for files.

Perhaps I could spare doing this at all though, so my change would be smaller.

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 10]
```diff
@@ -56,7 +56,6 @@
 import static com.teradata.benchto.driver.loader.BenchmarkDescriptor.VARIABLES_KEY;
```
Comment (sopel39) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Does it also work for classpaths?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/BenchmarkProperties.java [hunk 11]
```diff
@@ -31,17 +29,16 @@
...
 import static com.teradata.benchto.driver.utils.PropertiesUtils.splitProperty;
-import static com.teradata.benchto.driver.utils.ResourceUtils.asPath;
 import static java.util.stream.Collectors.toMap;
 
 @Component
 public class BenchmarkProperties
 {
 
-    @Value("${sql:sql}")
+    @Value("${sql:#{'classpath:sql'}}")
```
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/BenchmarkProperties.java]: 👎  broken

Comment (sopel39) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/BenchmarkProperties.java]: for `benchmarkDir` you use `classpath:/`,
here you use `classpath:`

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java [hunk 12]
```diff
@@ -139,32 +137,65 @@
         }
     }
 
-    private List<Path> findBenchmarkFiles()
+    private List<Resource> findBenchmarkFiles(ResourcePatternResolver patternResolver)
             throws IOException
     {
-        LOGGER.info("Searching for benchmarks in classpath ...");
+        String benchmarksDir = properties.getBenchmarksDir();
+        List<String> activeBenchmarkPatterns = activeBenchmarkPatterns();
```
Comment (sopel39) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Is it still a pattern? I though it's equality based.

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java [hunk 13]
```diff
@@ -122,6 +128,56 @@ public void shouldLoadSimpleBenchmark()
...
+        Set<String> loadedNames = benchmarks.stream()
+                .map(Benchmark::getName)
+                .collect(toSet());
+
+        assertThat(loadedNames)
+                .contains("concurrent-benchmark", "simple-benchmark");
+    }
+
+    @Test
+    public void shouldLoadBenchmarksFromFileSystem()
```
Comment (sopel39) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: Maybe add classpath test, e.g: `shouldLoadBenchmarksFromClasspath`? I am not sure if that is trivial though. What about multiple JARs?

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 2 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Reverted to make change smaller

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java (hunk 4 above)
Comment (findepi) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: If you're we live in a world where it's obvious what a test tests, then I am sure you're wrong :)
Requiring such comments serves as a reminder that code should be clear. Once it's clear, the comments are no longer needed. Up to you.

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 6 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Right.

When there is no `/` (or no leading `/`), the path is treated as relative to working directory.

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 8 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: There is not so much to be extracted.

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java (hunk 9 above)
Comment (findepi) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: We follow airlift code style, but it isn't checked. Out of scope.

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 10 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: > sans

http://dictionary.cambridge.org/dictionary/english/sans

Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: > Does it also work for classpaths?
  
Does, covered by existing tests

Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: > No tests needed to be changed? Maybe add sample test like the one in commit message.

That would only the purpose of cementing the design decision, without giving more coverage (function-wise).

File: benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java (hunk 13 above)
Comment (findepi) [inline benchto-driver/src/test/java/com/teradata/benchto/driver/loader/BenchmarkLoaderTest.java]: Tests load from classpath by default (see setup method)

File: benchto-driver/src/main/java/com/teradata/benchto/driver/BenchmarkProperties.java (hunk 11 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/BenchmarkProperties.java]: This is broken anyway

File: benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java (hunk 12 above)
Comment (findepi) [inline benchto-driver/src/main/java/com/teradata/benchto/driver/loader/BenchmarkLoader.java]: Changed. Done more like it used to be

**Comment by findepi** [comment]
@fiedukow suggested simplifying the driver and completely remove loading from classpath.

it was agreed offline
